jane/doc/proposals/unboxed-types/null.md

```diff
@@ -25,120 +25,110 @@ We thus want `t or_null` to be just like `t`, except now with a meaning
 environment:
 
 ```ocaml
-type ('a : non_null_value) or_null : value
+type ('a : value) or_null : value_or_null
 ```
 
-In addition, we will allow `'a or_null` to mode-cross the same way that
-`'a` does -- effectively allowing `or_null` to work nicely with `immediate`s.
-This can be implemented by "looking through" the
-`or_null` type constructor during mode-cross-checking. (The alternative would be to
-have some degree of kind polymorphism, where we have
-`type ('a : non_null_value & 'modes) or_null : value & 'modes`
-or similar, but looking through
-`or_null` is easy enough so that we can avoid this extra complexity.)
-This permits, e.g. `int or_null` to still be subject to
+With-kinds will allow `'a or_null` to mode-cross the same way that
+`'a` does -- this permits, e.g. `int or_null` to still be subject to
 optimizations around immediates.
 
-The `or_null` type has constructors `Some` and `None` (usable for both
+The `or_null` type has constructors `This` and `Null` (usable for both
 construction and pattern-matching), subject to the usual type-based
 disambiguation.
 
-## The choice of meaning of `value`.
+## Defaults
 
-In the diagram above, the `value` layout includes types that contain null,
-whereas the `non_null_value` layout is the one that includes all of today's
-types. This choice -- to make a *new* layout have the short name -- derives
-from our expectation that most annotations will prefer to denote the superlayout,
-not the sublayout. Here is an example:
+In the diagram above, the `value_or_null` layout includes types that do contain
+null, whereas the `value` layout is the one that includes all of today's
+types, and is the default for abstract types and type variables. Why?
 
-```ocaml
-type ('element : value, 'index : immediate) flexible_array
-```
-
-where `flexible_array` can work with a non-`int` index type. Such a storage
-structure likely does not care whether its elements support 0 or not. Yet, if we
-used `value` to mean the layout without null, then it would be impossible to store
-a `widget or_null` in the `flexible_array`. We thus believe that `or_null` will
-be easier to use if users' `value` accepts null -- even
-though our proposed `value` is distinct from the layout of legacy types.
-
-Users can always write the `non_null`
-versions if they choose to. Note that because of the way layout inference works,
-the choice between these layouts matters only for abstract types. For example,
-the following still works fine:
+It turns out that whatever default we choose for abstract types will also
+determine the default for type variables due to backwards compatibility.
+Consider the following examples:
 
 ```ocaml
-type ('a : value) pair_with_default = { contents : 'a or_null; default : 'a }
+module M : sig
+  type 'a t (* Assumed non-null. *)
+end = struct
+  type 'a t = 'a (* [’a] must be non-null. *)
+end
+
+module F (X : sig
+  type t (* Assumed maybe-null. *)
```

Comment:
Is this really true? I didn't know we had a different default in a functor sig. That seems hard.

Reply:
No, we don't. I was illustrating the reasoning why the defaults must agree. I will rewrite this to be clearer.